src/sentry/auth/elevated_mode.py

```diff
@@ -1,38 +1,47 @@
 from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Tuple
+
+
+class InactiveReason(str, Enum):
+    INVALID_IP = "invalid-ip"
+    INCOMPLETE_SSO = "incomplete-sso"
+    # Indicates the request should be allowed
+    NONE = None
+
+    def __bool__(self):
+        return self.value is not None
+
+    def __str__(self) -> str:
+        return self.value
 
 
 class ElevatedMode(ABC):
     @property
     @abstractmethod
-    def is_active(self):
+    def is_active(self) -> bool:
         pass
 
-    @classmethod
     @abstractmethod
-    def is_privileged_request(self):
+    def is_privileged_request(self) -> Tuple[bool, InactiveReason]:
         pass
 
-    @classmethod
     @abstractmethod
     def get_session_data(self, current_datetime=None):
         pass
 
-    @classmethod
     @abstractmethod
-    def _populate(self, current_datetime=None):
+    def _populate(self) -> None:
         pass
 
-    @classmethod
     @abstractmethod
-    def set_logged_in(self, user, current_datetime=None):
+    def set_logged_in(self, user, current_datetime=None) -> None:
```

Comment:
do we need the current time for this?

Reply:
It's used for testing in superuser which I copied over to staff